04-onion-routing.md

```diff
@@ -367,6 +381,239 @@ otherwise meets the amount criterion (eg. some other failure, or
+  - MUST decrypt it using `rho(i)` and `E(i)`
+  - MUST use the decrypted fields to locate the next node
+- If `encrypted_data` contains a `next_blinding_override`:
+  - MUST use it as the next blinding point
+- Otherise:
+  - MUST use `E(i+1)` as the next blinding point
+- MUST forward the onion and communicate the next blinding point to the next node
+
+The final recipient:
+
+- MUST compute:
+  - `ss(r) = SHA256(k(r) * E(r))` (standard ECDH)
+  - `b(r) = HMAC256("blinded_node_id", ss(r)) * k(r)`
+  - `rho(r) = HMAC256("rho", ss(r))`
+- If an `encrypted_data` field is provided:
+  - MUST decrypt it using `rho(r)` and `E(r)`
+- MUST ignore the message if `E(r)` or `path_id` does not match the
+  blinded route it created
+
+#### Rationale
+
+Route blinding is a lightweight technique to provide recipient anonymity.
+It's more flexible than rendezvous because it lets senders choose what data
+they include in each node's onion payload and retry with updated values if
+the route fails.
+
+Use-cases where route blinding is useful include:
+
+- Recipient anonymity when receiving payments
+- Using unannounced channels in invoices without revealing them
+- Forcing a message to go through a specific set of intermediaries that can witness it
+- Providing anonymous reply paths for onion messages (offers, stuckless payments, etc)
+
+Each node in the blinded route needs to receive `E(i)` to be able to decrypt
+the onion and the `encrypted_data`. Protocols that use route blinding must
+specify how this value is propagated between nodes.
+
+When concatenating two blinded routes generated by different nodes, the
+last node of the first route needs to know the first `blinding` of the second
+route: the `next_blinding_override` field must be used to transmit this data.
+
+The final recipient must verify that the blinded route is used in the right
+context (e.g. for a specific payment). It can do so by storing `E(r)` and the
+context (e.g. a `payment_hash`), and verifying that they match when receiving
+the onion. Otherwise, to avoid additional storage cost, it can put some context
+information in the `path_id` field (e.g. a `payment_hash`) and verify that when
+receiving the onion.
+
+The `padding` field can be used to ensure that all `encrypted_data` have the
+same length. It's particularly useful when adding dummy hops at the end of a
+blinded route, to prevent the sender from figuring out which node is the final
+recipient.
+
+
+### Onion Message Payload Format
+
+Onion messages have an onion with an alternate `hop_payload`
+format: a `bigsize` followed by a `onionmsg_payload`.  Note that there
+is no legacy format, thus a `bigsize` of 0 means no payload.
+
+An intermediate node expects an `encrypted_data_tlv` which it can
+decrypt using the `blinding` which it is handed along with the onion
+message.
+
+Field numbers 64 and above are reserved for payloads for the final
+hop.
+
+1. `tlv_stream`: `onionmsg_payload`
+2. types:
+    1. type: 2 (`reply_path`)
```

Comment:
How is this secure from spoofing the `reply_path`? 

As the originator from the Onion Message is not known couldn't one basically create a `reply_path` to any node in the network and create potential DoS attacks (especially if rate limiting of onion messages is being introduced)

Reply:
You would have to send one message to a random node so that your target is spammed with one message. How is that better than just sending the message to your target directly? You can indeed do that but it doesn't amplify the number of messages.

For a DoS attack you would more likely want to have as many hops as possible to amplify the attack but this reflection doesn't give you anything interesting.

Reply:
True, unless I find a way in which the node will send the reply via the `reply_path` several times because it thinks the reply has not been received as we use an unreliable protocol (this could eventually happen from application land)

Reply:
I suppose this will work. Though it seems still a bit risky